src/main/java/net/glowstone/block/data/AbstractBlockData.java

```diff
@@ -0,0 +1,86 @@
+package net.glowstone.block.data;
+
+import net.glowstone.block.data.state.StateGenerator;
+import net.glowstone.block.data.state.StateValue;
+import org.bukkit.Material;
+import org.bukkit.block.data.BlockData;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+
+public abstract class AbstractBlockData implements IBlockData {
+
+    private Material material;
+    private Set<StateValue<? extends Object>> values;
+
+    public AbstractBlockData(Material material, StateGenerator<? extends Object>... array){
+        this.material = material;
+        this.values = new HashSet<>(array.length);
+        for(StateGenerator<?> gen : array){
+            this.values.add(gen.getDefaultStateValue());
+        }
+    }
+
+    public AbstractBlockData(Material material, StateValue<? extends Object>... array){
+        this(material, Arrays.asList(array));
+    }
+
+    public AbstractBlockData(Material material){
+        this(material, new StateValue[0]);
+    }
+
+    public AbstractBlockData(Material material, Collection<StateValue<? extends Object>> collection){
+        this.material = material;
+        this.values = new HashSet<>(collection);
+    }
+
+    @Override
+    public Set<StateValue<?>> getStateValues(){
+        return this.values;
+    }
+
+    @Override
+    public @NotNull Material getMaterial() {
+        return this.material;
+    }
+
+    @Override
+    public @NotNull String getAsString(boolean b) {
+        return b ? "minecraft:" + this.material.name().toLowerCase() : getAsString();
+    }
+
+    @Override
+    public @NotNull String getAsString(){
+        String entries = null;
+        for(StateValue<?> value : this.values){
+            if(entries == null){
+                entries = value.getGenerator().getId() + "=" + value.getValueAsString();
+            }else{
+                entries = entries + ", " + value.getGenerator().getId() + ": " + value.getValueAsString();
```

Comment:
Did you mean to give the top line a different id-value separator from the bottom line?

Reply:
Nope. Haha. Thanks for noticing that